File: OJ practices/LeetCode题目集/23. Merge k Sorted Lists(hard)-建堆法.py
'''
@Autor: SourDumplings
@Date: 2020-01-25 22:01:49
@Link: https://github.com/SourDumplings/
@Email: changzheng300@foxmail.com
@Description: https://leetcode.com/problems/merge-k-sorted-lists/
'''

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
from queue import PriorityQueue
from typing import List


class Solution:
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        t = h = ListNode(-1)
        pq = PriorityQueue()
        for l in lists:
            while l is not None:
                pq.put(l.val)
                l = l.next
        
        while not pq.empty():
            t.next = ListNode(pq.get())
            t = t.next
        return h.next

File: OJ practices/LeetCode题目集/23. Merge k Sorted Lists(hard)-建堆法.java
import java.util.ArrayList;
import java.util.List;
import java.util.PriorityQueue;
import java.util.Queue;

/*
 * @Autor: SourDumplings
 * @Date: 2020-01-25 21:47:27
 * @Link: https://github.com/SourDumplings/
 * @Email: changzheng300@foxmail.com
 * @Description: https://leetcode.com/problems/merge-k-sorted-lists/
 */

/**
* Definition for singly-linked list.
* public class ListNode {
*     int val;
*     ListNode next;
*     ListNode(int x) { val = x; }
* }
*/
class Solution
{
    public ListNode mergeKLists(ListNode[] lists)
    {
        List<Integer> vals = new ArrayList<>();
        for (ListNode l : lists)
        {
            while (l != null)
            {
                vals.add(l.val);
                l = l.next;
            }
        }
        Queue<Integer> pq = new PriorityQueue<>(vals);
        ListNode h = new ListNode(-1);
        ListNode t = h;
        while (!pq.isEmpty())
        {
            int v = pq.poll();
            t.next = new ListNode(v);
            t = t.next;
        }
        return h.next;
    }
}